Common interface to interpret and execute API methods

rreusser opened the pull request:
This PR works toward a common interface for executing commands. This includes:

- [x] `Plots.executeAPICommand(gd, method, args)` which executes a command (i.e. from components)
- [x] `Plots.computeAPICommandBindings(gd, method, args)` which interprets the intent of a command and returns a list of properties which affect the value of the component.
- [x] cache and check values for bound values

Completion of this feature will complete the implementation of sliders (#986) with two-way binding and will be applicable to other components.

rreusser commented:
@etpinard @alexcjohnson The code isn't ready for review yet, but the internal usage is:

To hook this up, you provide it the gd, an array of commands, and a callback to make an update. If the commands are straightforward, it will create lookup table by value so that when a restyle or relayout comes through, it can easily check to see if the value has changed (via _fullLayout or _fullData) and execute the callback if there appears to be a matching component state.

From `sliders`:

```javascript
        if(!sliderOpts._commandObserver) {
            sliderOpts._commandObserver = Plots.createCommandObserver(gd, sliderOpts.steps, function(data) {
                setActive(gd, d3.select(this), sliderOpts, data.index, false, true);
            }.bind(this));
        }  
```

rreusser commented:
(oh, and there's also a `sliderOpts._commandObserver.remove()` that needs to be called on d3 exit)

etpinard commented on src/plot_api/plot_api.js:
```diff
@@ -2124,6 +2126,7 @@ Plotly.update = function update(gd, traceUpdate, layoutUpdate, traces) {
             data: restyleSpecs.eventData,
             layout: relayoutSpecs.eventData
         });
+        gd.emit('plotly_plotmodified');
```
why not just subscribe to the existing [`plotly_afterplot`](https://github.com/plotly/plotly.js/blob/95c56e840aecf0c0a8a99a71341895f6bacbe4ab/src/plot_api/plot_api.js#L327) ?

rreusser replied:
Didn't know about it! Just need to add that event in a couple more places, but that should be perfect.

rreusser replied:
Ah, other reason, now that I think about it:

Should afterplot be emitted between frames? The exact sequence definitely matters, so the other logic was that a unique event would free us from tight coupling.

etpinard replied:
> Should afterplot be emitted between frames?

Hmm. Very good question. I'd say _no_. `plotly_afterplot` should be the fired just before the promise is returned. But, I can be convinced otherwise.

rreusser replied:
Do you think it should be a separate event for this purpose? Again, my fear was coupling the component update code with events that serve other purposes. But I can try to use `plotly_afterplot` if you prefer. (though, again, it would need to be called during animation)

etpinard replied:
I have no strong opinion here. I'll leave it up to you.

rreusser commented on src/plots/command.js:
```diff
@@ -0,0 +1,369 @@
+            'plotly_update',
+            'plotly_animatingframe',
+            'plotly_afterplot'
+        ];
```
@etpinard I refactored this to just hook into a list of existing events. This seems much simpler and more straightforward.

etpinard replied:
Very nicely done. Thank you

rreusser commented:
I think this is ready for review. I'm going to go back over the code, but here's a quick summary of need-to-know:

- `Plots.hasSimpleAPICommandBindings` looks over a `method` and `args` to see if it's simple enough that we can infer which state is active (currently limited to one property but maybe multiple traces). When run. That means it looks over the *whole* list of component commands, so it caches a lookup table of value -> state mappings as it goes and bails out as soon as the answer is that it's not simple.

- `Plots.createCommandObserver`: hooks into events that could conceivably change things. Each time it's called, it compares the current value to the cached value. If things have changed, it executes a callback with the index from the lookup table computed in the first step. Needs to be cleaned up with `remove()` so that event listeners are properly removed.

- `Plots.executeAPICommand`: a common interface for executing API commands. There's hardly anything to it, but seems to make sense to put this in one place.

- `Plots.computeAPICommandBindings`: Exposed for convenience, but currently only used via `Plots.hasSimpleAPICommandBindings`. This is the magic that takes a method args and computes the bindings for it.

cc: @etpinard @alexcjohnson

rreusser commented:
Also, here's a live illustrating what we're actually talking about here: http://rickyreusser.com/animation-experiments/#binding

etpinard commented on test/jasmine/tests/binding_test.js:
```diff
@@ -0,0 +1,27 @@
+    });
+
+    it('updates bindings on plot', function(done) {
+        Plotly.restyle(gd, 'marker.size', 10).catch(fail).then(done);
```
Would you mind adding an asserting checking if the updatemenus DOM nodes updated correctly. This might be easier to check using when restyling `'marker.color'` instead of `'marker.size'`.

etpinard commented on test/jasmine/tests/command_test.js:
```diff
@@ -0,0 +1,477 @@
+    describe('restyle', function() {
+        describe('with invalid notation', function() {
+            it('with a scalar value', function() {
+                var result = Plots.computeAPICommandBindings(gd, 'restyle', [['x']]);
```
Thanks for checking this here - as we don't currently don't dive into `args` checking that its items are valid :tada:

etpinard commented on test/jasmine/tests/command_test.js:
```diff
@@ -0,0 +1,477 @@
+                    y: [[3, 4, 5]],
+                    'marker.size': [10, 20, 25],
+                    'line.color': 'red',
+                    line: {
```
So, `line: { width: [2, 8] }` does not override `line.color: 'red'` here. Go figure. Thanks for checking!

rreusser replied:
Oh no! Good catch. I changed the order so that this is no longer accurate 😞 I need to fix this. Sigh.

etpinard replied:
> otherwise I'm fine documenting it and moving on

I'm fine with that also.

etpinard commented on test/jasmine/tests/command_test.js:
```diff
@@ -0,0 +1,477 @@
+    });
+});
+
+describe('component bindings', function() {
```
How is this suite different than the new `binding_test.js` here:

https://github.com/plotly/plotly.js/pull/1016/files#diff-7f09c177b80364e96a43622b17a9d8e7R24

?

etpinard replied:
Oh I see, one test sliders and the other updatemenus.

@rreusser would you mind merging `binding_test.js` into this `describe` block?

rreusser replied:
Moved.

rreusser commented on src/plots/command.js:
```diff
@@ -0,0 +1,369 @@
+
+    // Determine whether there's anything to do for this binding:
+    var binding;
+    if((binding = exports.hasSimpleAPICommandBindings(gd, commandList, lookupTable))) {
```
This kinda became a degenerate case where I might as well move it into the above variable declaration. It's nothing fancy. Just assigns a variable and checks it in one step.

etpinard replied:
got it. thanks!

rreusser commented on src/plots/command.js:
```diff
@@ -0,0 +1,369 @@
+        allArgs.push(args[i]);
+    }
+
+    if(!apiMethod) {
```
Correct. I have no strong feelings either way; it's just a general habit analogous to `callback && callback()` in order to ensure I never unexpectedly invoke `undefined()`. Thought maybe safety for depending on how it gets used in the future, but I'm fine just removing. I'm not great at proper promise error handling. Few are.

rreusser commented on src/plots/command.js:
```diff
@@ -0,0 +1,369 @@
+        return Promise.reject();
+    }
+
+    return apiMethod.apply(null, allArgs);
```
Mixed feelings. Ditto above on really struggling with *proper* error handling with promises.

etpinard replied:
I have no strong feelings about this. I'll let you decide what's best. :+1:

rreusser commented on src/plots/command.js:
```diff
@@ -0,0 +1,369 @@
+        case 'animate':
+            bindings = computeAnimateBindings(gd, args);
+            break;
+        default:
```
I guess really it should `throw` so that when we add api methods and someone forgets to whitelist them here, they'll get a hard error.

rreusser replied:
Now throws error.

etpinard commented on src/plot_api/plot_api.js:
```diff
@@ -2324,6 +2317,15 @@ Plotly.animate = function(gd, frameOrGroupNameOrFrameList, animationOpts) {
                     newFrame.transitionOpts
                 );
+
+                gd.emit('plotly_animatingframe', {
```
Thanks for the info :+1:

rreusser commented on src/plots/command.js:
```diff
@@ -0,0 +1,369 @@
+    var container, value, obj;
+    var changed = false;
+
+    if(binding.type === 'data') {
```
Good observation. Had to change the command a little, but dried it up 👍

rreusser commented on test/jasmine/tests/command_test.js:
```diff
@@ -0,0 +1,641 @@
+        destroyGraphDiv(gd);
+    });
+
+    it('attaches and updates bindings for sliders', function(done) {
```
@etpinard See tests. It checks:

- event listeners are attached when sliders are added
- event listeners are removed when sliders are update and are no longer bindable
- changing slider options triggers recomputing the lookup table so it still works

rreusser commented:
@etpinard I think this is reviewable. I've addressed the concerns and have added a pretty detailed test for both updatemenus and sliders that tests values changing.

Unfortunately (since this isn't immutablejs), detecting arg changes means we need to scan through the steps more often than I'd like (on each relayout) since we can't otherwise be certain whether the values have changed, but it's a pretty cheap check and tries to bail out early and cache when possible. SVG rendering should almost always be way more of a bottleneck than iterating through a few values, so I'm okay with it.

etpinard commented on src/components/sliders/draw.js:
```diff
@@ -71,14 +71,12 @@ module.exports = function draw(gd) {
-            sliderOpts._commandObserver = Plots.createCommandObserver(gd, sliderOpts.steps, function(data) {
-                if(sliderOpts.active === data.index) return;
-                if(sliderOpts._dragging) return;
+        Plots.createCommandObserver(gd, sliderOpts, sliderOpts.steps, function(data) {
```
I'm big a fan of this new pattern here where one routine creates + updates command observer. This should make the command observer logic more easily reused in the future when we integrate it in other components :+1: 

One small suggestion, now that `createCommandObsever` can create and update a command observer, maybe we should rename it something like `manageCommandObserver` similar to [`manageModebar`](https://github.com/plotly/plotly.js/blob/ec0b0a49326f689dee01b228704dfb9ac1c9ea78/src/components/modebar/manage.js#L27)?

rreusser replied:
Sounds good. I was going to go with `createOrUpdateCommandObserver`, but that seemed excessive. Changing now.

etpinard commented on test/jasmine/tests/command_test.js:
```diff
@@ -328,7 +328,15 @@ describe('Plots.computeAPICommandBindings', function() {
+                    // objects is not strictly defined but is at least consistent across browsers.
+                    // The worst-case scenario right now isn't too bad though since it's an obscure
+                    // case that will definitely cause bailout anyway before any bindings would
+                    // happen.
```
thanks !

etpinard commented on src/components/updatemenus/defaults.js:
```diff
@@ -33,6 +33,8 @@ module.exports = function updateMenusDefaults(layoutIn, layoutOut) {
         // used to determine object constancy
         menuOut._index = i;
 
+        menuOut._input.active = menuOut.active;
```
Hmm. Why do we need this?

etpinard commented on src/components/updatemenus/defaults.js:
```diff
@@ -48,7 +50,9 @@ function menuDefaults(menuIn, menuOut, layoutOut) {
     if(!visible) return;
 
-    coerce('active');
+    // Default to zero since active must be *something*:
```
If that's the case, we should add a `dflt` in the attribute declaration.

rreusser replied:
Ah, good point. That works for me. I think that's best since `undefined` isn't a state that really makes sense (because it's always *something*).

rreusser replied:
Oops. The default is set. Removing this. I think this may have been a change that I thought was necessary while debugging but ultimately did not need.

rreusser replied:
Reverted updatemenus change and fixed sliders defaults.

etpinard commented on test/jasmine/tests/command_test.js:
```diff
@@ -0,0 +1,641 @@
+    it('udpates bound components when the value changes', function(done) {
+        expect(gd.layout.sliders[0].active).toBe(0);
+
+        Plotly.restyle(gd, 'marker.color', 'blue').then(function() {
```
I'm a big fan of this test and the next. :beers:

rreusser replied:
It's pretty deluxe.

rreusser replied:
Ah, confused it with the deluxe tests down below. Yeah, these are just nice and simple 😄

etpinard commented on test/jasmine/tests/updatemenus_test.js:
```diff
@@ -225,6 +225,33 @@ describe('update menus buttons', function() {
+    });
+
+    it('does not set active on initial plot', function() {
+        expect(gd.layout.updatemenus[0].active).toBeUndefined();
```
very nice. Thanks for :lock: this !

etpinard commented:
Hard fought :dancer: 

Awesome work @rreusser
